### ISSUE TITLE
Java feature server fails to retrieve features from multiple feature views

### ISSUE THREAD

**Issue by vstrimaitis** [feast-dev/feast#2302]
## Expected Behavior 
When querying features from multiple feature views, the server should be able to fetch the values of all features successfully.

## Current Behavior
The server responds with `NOT_FOUND` for all features if either of these conditions is true:

* The `features` section of the request contains at least two features from different feature views
* The `entities` section of the request contains at least one excess entity

## Steps to reproduce
* Define two feature views with different entities
* Materialize the data to Redis
* Start the Java feature server
* Send a query asking to fetch features from different feature views. Example:

```json
{
    "features": {
        "val": [
            "feature_view_1:feature_1",
            "feature_view_1:feature_2",
            "feature_view_2:feature_3"
        ]
    },
    "entities": {
        "entity_1_id": {
            "val": [
                {
                    "int32_val": 1
                },
                {
                    "int32_val": 2
                }
            ]
        },
        "entity_2_id": {
            "val": [
                {
                    "int32_val": 10
                },
                {
                    "int32_val": 11
                }
            ]
        }
    }
}
```

The server will respond with statuses `NOT_FOUND` for all features. If you were to remove `"feature_view_2:feature_3"` and `"entity_2_id"` the server should respond with actual values. 

### Specifications

- Version: 0.18.0
- Platform: Linux

## Possible Solution
I did some investigation and compared the functionality of the Python and Java feature server implementations. The Python server seems to split the requested features by feature view and then make separate requests for each of them. The Java server doesn't do that and that probably messes up the Redis key generation down the line.

I was able to implement this grouping by feature view locally and everything seems to be working now. If you feel like this is the correct approach, I would be happy to clean my code up and send in a PR!

**Comment by woop** [comment]
One problem with this approach is that pre-emptive splitting makes it impossible for the connector to optimize lookups. For example if we are storing multiple feature views on a single entity key (as is the case with the Redis HMGET implementation), then we will force the connector to duplicate lookups which will increase latency. This is a shortcoming in the Python implementation. Alternatively, we could decide to just give up on that optimization in favor of simplicity.

**Comment by felixwang9817** [comment]
@woop if we introduced the batch retrieval interface to online stores (i.e. get multiple FVs, not just a single FV), couldn't we just pass the read through to the online store, which can make individual optimizations as necessary? I think this would allow the public interface to remain simple.

**Comment by vstrimaitis** [comment]
I submitted my current fix for the issue in PR #2316. My approach was basically to collect all unique entity combinations for the queried feature views and split the queries according to these combinations. If I understand everything correctly, there shouldn't be duplicate lookups, but I might be wrong. Let me know if the approach is not feasible or if there's something that can be fixed/improved 😁

**Comment by felixwang9817** [comment]
thanks for the PR @vstrimaitis, I'll take a look at it soon!
